source/extensions/http/header_validators/envoy_default/http1_header_validator.cc

```diff
@@ -0,0 +1,416 @@
+  } else if (key_string_view.at(0) != ':') {
+    // Validate the generic header name.
+    auto name_result = validateGenericHeaderName(key);
+    if (!name_result) {
+      return name_result;
+    }
+  } else {
+    // The only valid pseudo header for responses is :status. If the header name starts with ":"
+    // and it's not ":status", then the header name is an unknown pseudo header.
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidPseudoHeader};
+  }
+
+  // Validate the header value
+  return validateGenericHeaderValue(value);
+}
+
+::Envoy::Http::HeaderValidator::RequestHeaderMapValidationResult
+Http1HeaderValidator::validateRequestHeaderMap(RequestHeaderMap& header_map) {
+  absl::string_view path = header_map.getPathValue();
+  absl::string_view host = header_map.getHostValue();
+  // Step 1: verify that required pseudo headers are present. HTTP/1.1 requests requires the
+  // :method header based on RFC 9112
+  // https://www.rfc-editor.org/rfc/rfc9112.html#section-3:
+  //
+  // request-line   = method SP request-target SP HTTP-version CRLF
+  //
+  // The request-target will be stored in :path except for CONNECT requests which store the
+  // request-target in :authority. So we only check that :method is set initially.
+  if (header_map.getMethodValue().empty()) {
+    return {RequestHeaderMapValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().InvalidMethod};
+  }
+
+  // HTTP/1.1 also requires the Host header,
+  // https://www.rfc-editor.org/rfc/rfc9112.html#section-3.2:
+  //
+  // A client MUST send a Host header field in all HTTP/1.1 request messages.
+  // ...
+  // A client MUST send a Host header field in an HTTP/1.1 request even if the
+  // request-target is in the absolute-form
+  // ...
+  // If the authority component is missing or undefined for the target URI, then a
+  // client MUST send a Host header field with an empty field-value.
+  if (host.empty()) {
+    return {RequestHeaderMapValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().InvalidHost};
+  }
+
+  // Verify that the path and Host/:authority header matches based on the method.
+  // From RFC 9112, https://www.rfc-editor.org/rfc/rfc9112.html#section-3.2.2:
+  //
+  // When a proxy receives a request with an absolute-form of request-target, the
+  // proxy MUST ignore the received Host header field (if any) and instead replace
+  // it with the host information of the request-target. A proxy that forwards
+  // such a request MUST generate a new Host field-value based on the received
+  // request-target rather than forward the received Host field-value.
+  // ...
+  // If the target URI includes an authority component, then a client MUST send a
+  // field-value for Host that is identical to that authority component,
+  // excluding any userinfo subcomponent and its "@" delimiter (Section 2.7.1).
+  //
+  // TODO(meilya) - should this be implemented here in UHV or the H1 codec?
```

Comment:
I will hold off on actually performing this check until we have path normalization because we need to know the path form prior to performing this check: verify that the `:path` is in authority-form, parse the authority portion, compare the path authority portion against the `:authority` header.

Reply:
I will hold off on actually performing this check until we have path normalization because we need to know the path form prior to performing this check: verify that the `:path` is in authority-form, parse the authority portion, compare the path authority portion against the `:authority` header.